src/freelunch/optimisers.py

```diff
@@ -1,145 +1,402 @@
+        Khat_best = (herd_best[0] - Kfood) / spread
+
+        # Beta best
+        beta_best = Khat_best[:,None]*Xhat_best
+    
+        beta = beta_food + beta_best
+
+        # Foraging motion
+        inertia = tech.lin_reduce(self.hypers['Iforage'],gen,self.hypers['G']) 
+        return self.hypers['Vf']*beta + inertia*herd[3]
+
+
+    def random_diffusion(self,gen):
+
+        delta = 2*np.random.rand(self.hypers['N'],self.obj.n) - 1
+
+        return self.hypers['Dmax']*( 1 - gen/self.hypers['G'])*delta
+        
+
+
+    def run(self):
+
+        pop = self.init_pop(self.hypers['N'])
+
+        # Compute first set of fitness
+        pop = tech.compute_obj(pop, self.obj)
+  
+
+        # Determine time step as in paper
+        if self.bounds is None:
+            dt = 10 # If no bounds set use default
+        else:
+            bounds = tech.bounds_as_mat(self.bounds)
+            dt = self.hypers['Ct']*np.sum(bounds[:,1]-bounds[:,0])
+
+
+        # Main loop 
+        for gen in range(self.hypers['G']):
+
+            herd = self.get_herd(pop)
+
+            # Induced motion, following the crowd
+            N, Khat_best = self.local_motion(herd,gen)
+
+            # Foraging motion
+            F = self.foraging(herd,gen,pop)
+
+            # Random diffusion
+            D = self.random_diffusion(gen)
+
+            # Total Velocity
+            V = N + F + D
+
+            # Genetic operations
+            champion = self.all_time_champion(pop)
+
+            # Crossover
+            if self.hypers['Crossover']:
+                crossover_prob = 0.2*Khat_best
+                xover_herd = herd[1][np.random.randint(self.hypers['N'],size=(self.hypers['N'],)),:]
+                for i,h in enumerate(herd[1]):
+                    # Implement 1-to-gbest X-over not 1-to-rand as in paper...
+                    xover_herd[i,:] = tech.binary_crossover(h,champion[1],crossover_prob[i])
+                current_herd = xover_herd
+            else:
+                current_herd = herd[1]
+
+                
+            # Mutation
+            if self.hypers['Mutate']:
+                mutate_prob = 0.05/Khat_best
+                mutate_prob[mutate_prob==np.Inf] = 0
+                inds = np.random.randint(self.hypers['N'],size=(self.hypers['N'],2))
+                mutates = np.random.rand(self.hypers['N'],self.obj.n) < mutate_prob[:,None]
+                mutant_dna = champion[1] + self.hypers['Mu']*(herd[1][inds[:,0],:]-herd[1][inds[:,1],:])
+                current_herd[mutates] = mutant_dna[mutates]
+
+            # Move the herd
+            new_pos = current_herd + dt*V
+
+            # Compute objectives and update the herd
+            for i,(dna,motion,forage) in enumerate(zip(new_pos,N,F)):
+                pop[i].pos = dna
+                pop[i].motion = motion
+                pop[i].forage = forage
+                pop[i].fitness = self.obj(dna)
+
+        return [ p.as_sol() for p in pop ]

```

Comment:
does this return the solutions in fitness order?  I think the API assumes that they are, I'll double check and add a test case for this.

Reply:
Doesn't currently return in order. My bad. Also not in PSO, will open issue.

Reply:
I think the build will fail until then so will hold off on merging the PR for now